lib/typed/failure.rb

```diff
@@ -43,9 +43,29 @@ def failure?
       true
     end
 
-    sig { override.returns(NilClass) }
+    sig { override.returns(T.noreturn) }
     def payload
-      nil
+      raise "No payload on a failed result."
```

Comment:
Do we want a custom error type here? Trying to think through if `StandardError` is sufficient, or maybe `NotImplemented` or `Argument` would work better?

Reply:
Done 👍 !